### PR TITLE
HDDS-9444. S3SecretManager fails to revoke secret due to inconsistency between S3Cache and S3SecretTable.

### PR DESCRIPTION
**Pull request by ArafatKhan2198**
### Issue Summary:

The testGetSetRevokeS3Secret method in the TestSecureOzoneCluster class intermittently fails when testing the revocation of Ozone S3 secrets. The problem lies in how the cache handles revoked secrets, leading to unexpected behavior.

### Details of the Problem:

1. The test method creates an access ID, retrieves an S3 secret, revokes it, and tries to set a new one. The last step should fail after revocation but sometimes succeeds, allowing access and modification of the secret key.

2. Investigation revealed that the cache removes the entire access ID and secret key record upon revocation, causing confusion between revoked and active secrets.

3. The code checks the cache, returning null due to revocation, but later successfully retrieves the secret from the DB without the expected error due to cache-DB inconsistency.

The lag in updating the DB is due to the Flush Thread not promptly committing cache changes.

### Proposed Solution:

- To address this issue, we propose retaining the access ID record in the cache after revocation. Additionally, we'll introduce a boolean isDeleted; identifier in S3SecretValue set to true for revoked secrets. This way, accessing or setting revoked secrets will prompt cache notifications, avoiding unnecessary DB checks when the DB is not up-to-date.
- We will also store the transaction IDs of the transactions in the S3SecretValue instance. This is necessary as we need to track the transactions in the cache that have been flushed by the doubleBuffer and must subsequently remove them from the cache. The transaction ID that is written to the DoubleBuffer must also be stored in the S3SecretValue.

CI Run :- https://github.com/ArafatKhan2198/ozone/actions/runs/6468031778/job/17559261272*#step:5:3776

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9444

## How was this patch tested?
Added more unit tests to verify the solution.

### CODE REVIEW

**Comment by ArafatKhan2198** [comment]
@ashishkumar50 @sumitagrawl @sadanand48 Can you please take a look!

**Comment by adoroszlai** [comment]
@Pochatkin @Tejaskriya please take a look

**Comment by sadanand48** [comment]
Thanks @ArafatKhan2198 for the patch & @Pochatkin for the comment.
I think we should make the cache unbounded i.e it should not invalidate an entry based on size or expiry time . The invalidation should only happen during OM double buffer flush. I'm not exactly sure about the rationale behind using guava cache, Using RocksDb Table cache could have avoided these issues .
@ArafatKhan2198  We should add the logic to invalidate cache (delete entries) during Double Buffer flush which is not there currently.

**Comment by sadanand48** [comment]
Thanks @Pochatkin for explaining, The current problem around getting a revoked key from the underlying store exists for both RocksDB and vault implementations . Making the cache unbounded would solve this, The eviction can only happen during doublebuffer flush.

**Comment by ArafatKhan2198** [comment]
@Pochatkin @sadanand48 thanks for the comments I have mentioned out the problem in a detailed manner along with the proposed fix :- 
### Cause of the problem ➖

- TestSecureOzoneCluster ➖
    - This test class has a method which is failing intermittently, called as `testGetSetRevokeS3Secret` which tests out the revoke feature of Ozone S3 secrets for a certain user. 
    https://github.com/apache/ozone/blob/377d6a674f76805d6ed4df640a87159757ed3398/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestSecureOzoneCluster.java#L698
    - The code flow for the test method and the problem encountered is going like this ➖
        1. A user accessID is created 
        2. It attempts to get an S3 secret for the user accessID. 
        3. It revokes the existing S3 secret.
        4. It attempts to `set` a new S3 secret for the user, which should fail and throw an error *`ACCESS_ID_NOT_FOUND`* because the access ID (username) has been revoked earlier for that user. In most of the cases it throws an error but in a few of them it does not because somehow even though the user has been revoked it is still somehow able to access the secret key for that user and change it. 
        5. The reason for that was investigated by adding logs to the method called as S3SecretManager method `getSecret` which is responsible for checking both the cache and the DB for the secret the user requires. https://github.com/apache/ozone/blob/377d6a674f76805d6ed4df640a87159757ed3398/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/S3SecretManagerImpl.java#L54
        6. Now when we tried to set a new s3 secret for an earlier revoked user. It first proceeded to check the cache but it could find one and returned null because the earlier revoke method had invalidated the user accessID and S3Secret from the cache hearby completely removing it.➖
            
            `S3SecretValue cacheValue = s3SecretCache.get(kerberosID);`
            
        7. Afterwards it decided to check the DB S3SecretKeyTable for the secret for the user and for some reason it was able to fetch one and returned it back and was later changed without throwing an error *`ACCESS_ID_NOT_FOUND`* like it shoul. The problem here happened was that the DB was not informed about the changes in cache that had happened yet, hence this presented an inconsistency between the cache and DB. 
        `S3SecretValue result = s3SecretStore.getSecret(kerberosID);`
        8. The reason for that lag was due to the fact that the background Flush Thread had not yet commited the changes of the Cache to the DB. This was confirmed by the logs I added where we checked the cache and the db for the revoked secret key for the user. 
             ```
               09:44:46,260 INFO  om.S3SecretManagerImpl (S3SecretManagerImpl.java:getSecret(59)) - Get secret for kerberosID:om/@EXAMPLE.COM from cache:null
               09:44:46,260 INFO  om.S3SecretManagerImpl (S3SecretManagerImpl.java:getSecret(66)) - Get secret for kerberosID:om/@EXAMPLE.COM from store:ae367331576fa164eaa43fe906014b579e2df3698ef4c744c5bdc88c8dec538f
             ```

### Fix for the Problem ➖

- Previously, in the earlier RocksDB cache implementation, when a secret was revoked, the system retained the access ID record, but the respective secret was stored as null. This null value served as an indicator for the SecretManager that a specific access ID had been revoked, eliminating the need to check the database for that user. However, with the implementation of a newer cache using the Guava Google library, the entire record of the access ID and the secret key is removed from the cache upon revocation. This change poses a challenge for the SecretManager because it can no longer distinguish whether a requested secret has been revoked or not.
- To address this issue, we propose a fix that enables the SecretManager to maintain the record even after revocation. Additionally, an identifier, **`boolean isDeleted;`**, will be introduced in the S3SecretValue. This identifier will be set to `true` whenever a secret has been revoked, and the value of the secret itself will be set to null. Consequently, in the future, if we attempt to access or set a revoked secret, the cache will promptly notify us that the secret has been revoked. This way, we can avoid unnecessary database checks, especially when the database has not been updated with the latest data.

**Comment by ChenSammi** [comment]
For this specific problem, the proposed solution by @ArafatKhan2198 plus an unlimited size Guava cache will solve the problem. 
But besides this one, I see there were other issues caused by inconsistency between TableCache and DB.  So if we can make the DB consistent with TableCache after each call returns, then we can resolve all these kinds of issues caused by this inconsistency, once for all. What do you think @sadanand48 ?

**Comment by sadanand48** [comment]
Yes @ChenSammi , The inconsistency problem is solved in the current patch by adding a delete marker in the cache but since cache has a bound, it can cause issues. Like I mentioned above , The changes that need to be made in this PR are
- make guava cache unbounded . This can be simply done by not setting maximumSize and expiryTime in the cache builder as by default behaviour is unbounded
  ``` java   
  cache = CacheBuilder.newBuilder()
          .expireAfterWrite(expireTime)
          .maximumSize(maxSize)
          .build();
  ```
- During double buffer flush the cache needs to be invalidated

**Comment by ChenSammi** [comment]
@sadanand48 , I maybe not explain it clearly. My alternative suggestion is can we remove the lag caused by double buffer flush? If every time after a user request is acked by OM, we can assume visit either TableCache or DB will get the same result, then there will be no inconsistency problem, not only for this S3 secret table, but for all other tables which are partially cached.  If this works, then I will prefer this solution.

**Comment by ArafatKhan2198** [comment]
@ChenSammi  The delay was due to the background Flush Thread not having committed the cache changes to the DB yet. This is expected, as there's a set time interval for flushing the data. Are you suggesting we skip the DB check for the revoked secret and only verify it after a flush is initiated?
Additionally, the S3 secret could not get revoked because `S3SecretManager` thought that since the secret is not present in the cache hence we should check the table where it found it. but in reality the user had revoked it previously but `S3SecretManager` had no idea of knowing that since the entire entry was removed from the cache with no info about the previous revoked attempt. So, there's a mix-up, but it's more like confusion rather than an Inconsistency problem.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/S3InMemoryCache.java [hunk 1]
```diff
@@ -60,23 +71,8 @@ public S3SecretValue get(String id) {
    * Builder for {@link S3InMemoryCache}.
    */
   public static class S3InMemoryCacheBuilder {
```
Comment (tanvipenumudy) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/S3InMemoryCache.java:73]: Thank you @ArafatKhan2198 for the patch. 

We can remove this builder: `S3InMemoryCache`.

**Comment by ChenSammi** [comment]
@sadanand48 , thanks for the explain. This TableCache usage is a bit different than what I thought.  The current proposal is good for it.

**Comment by adoroszlai** [comment]
@ArafatKhan2198 please check compile error:

https://github.com/ArafatKhan2198/ozone/actions/runs/6665326313/job/18114677697#step:6:1860

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java [hunk 2]
```diff
@@ -285,6 +295,8 @@ private void addToBatchTransactionInfoWithTrace(String parentName,
   void flushTransactions() {
     while (isRunning.get() && canFlush()) {
       flushCurrentBuffer();
+      // Invalidate all the content in S3SecretManager cache.
+      s3SecretManager.clearS3Cache();
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java:299]: We cannot just simply cleanup all entries at once, only the committed transactions need to be cleaned up. RocksDb table cache uses trxLogIndex as an epoch for the cache entry. 
Refer to cleanup cache method in this class. The idea is to track the committed trxs and only remove those from the cache.

Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java:299]: Thanks for the comment @sadanand48 

I believe that the cache entries are always cleared during each double buffer flush, which means that eventually, everything must be written to the database at some point in time, correct? Or is there a possibility that something might be left behind in the cache? 
Please correct me if I am going wrong somewhere!

Comment (sadanand48) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java:299]:
```java
      buffer.iterator().forEachRemaining(
          entry -> addCleanupEntry(entry, cleanupEpochs));
```
Let's say there were entries added to the buffer after this line, In that case we cannot cleanup all entries in cache as cache might contain entries that have not yet been considered for flush in the batch. To solve this we maintain cleanup epochs and only delete those particular entries for cleanup from cache that have been flushed.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestSecureOzoneCluster.java [hunk 3]
```diff
@@ -1414,7 +1414,7 @@ private static X509CertificateHolder generateX509CertHolder(
         .setSubject(subject)
         .setDigitalSignature(true)
         .setDigitalEncryption(true);
-    
+
```
Comment (ChenSammi) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestSecureOzoneCluster.java:1417]: Please revert this unnecessary change, which is the only change of this file.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java [hunk 4]
```diff
@@ -732,12 +735,14 @@ public OpenKeySession openKey(OmKeyArgs args) throws IOException {
 
   private OMResponse handleError(OMResponse resp) throws OMException {
     if (resp.getStatus() != OK) {
+      LOG.error("OMException occurred: {}", resp.getMessage());
```
Comment (ChenSammi) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java:738]: I would suggest remove this log, for every failure should be specifically handled and logged by where it's called in client.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/S3SecretManagerImpl.java [hunk 5]
```diff
@@ -118,4 +132,19 @@ public S3SecretCache cache() {
   public S3Batcher batcher() {
     return s3SecretStore.batcher();
   }
+
+  public void updateCache(String kerberosID, S3SecretValue secret) {
```
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/S3SecretManagerImpl.java:136]: add @Override

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/S3SecretManagerImpl.java [hunk 6]
```diff
@@ -118,4 +132,19 @@ public S3SecretCache cache() {
...
+    S3SecretManager.super.updateCache(kerberosID, secret);
+    // Mark the Kerberos ID as updated.
+    updatedKerberosIDs.add(kerberosID);
+  }
+
+  public void clearCache() {
```
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/S3SecretManagerImpl.java:142]: add @OverRide

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java [hunk 7]
```diff
@@ -194,7 +202,8 @@ private OzoneManagerDoubleBuffer(OMMetadataManager omMetadataManager,
       OzoneManagerRatisSnapshot ozoneManagerRatisSnapShot,
       boolean isRatisEnabled, boolean isTracingEnabled,
       Function<Long, Long> indexToTerm, int maxUnFlushedTransactions,
-      FlushNotifier flushNotifier, String threadPrefix) {
+      FlushNotifier flushNotifier, S3SecretManager s3SecretManager,
+                                   String threadPrefix) {
```
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java:206]: Please fix the indent.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/S3SecretManagerImpl.java [hunk 8]
```diff
@@ -128,4 +132,19 @@ public S3SecretCache cache() {
...
+    // Remove Kerberos IDs that have been updated from the cache.
+    for (String kerberosID : updatedKerberosIDs) {
+      S3SecretManager.super.invalidateCacheEntry(kerberosID);
```
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/S3SecretManagerImpl.java:145]: There are issues here. Invalidate action of an element, previously the element will be removed from the cache, currently the element will be marked as deleted, which is the secret revoked. So here, mark all accessID as deleted/revoked, is not what we want to do. 

Second, we cannot remove all ID in the updatedKerberosIDs without the epoch info, for some of the IDs' transaction are not flushed to rocksdb yet.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/S3InMemoryCache.java [hunk 9]
```diff
@@ -49,10 +51,38 @@ public void invalidate(String id) {
...
+   * @param flushedTransactionIds A list of transaction IDs that have been
+   *                              flushed and should be used to identify and
+   *                              remove corresponding cache entries.
    */
-  public void clearCache() {
-    cache.invalidateAll();
+  @Override
+  public void clearCache(List<Long> flushedTransactionIds) {
+    // Create a map to store transactionLogIndex-to-cacheKey mappings
+    Map<Long, Set<String>> transactionIdToCacheKeys = new HashMap<>();
```
Comment (sadanand48) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/S3InMemoryCache.java:68]: Can one transactionLogIdx be mapped to multiple cache keys ? I don't think so.

Comment (ArafatKhan2198) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/S3InMemoryCache.java:68]: Yes you are right every Ratis transaction will have a different index. I'll change that!
Thanks @sadanand48

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java [hunk 10]
```diff
@@ -732,12 +735,14 @@ public OpenKeySession openKey(OmKeyArgs args) throws IOException {
...
       throw new OMException(resp.getMessage(),
           ResultCodes.values()[resp.getStatus().ordinal()]);
     }
     return resp;
   }
 
+
```
Comment (sadanand48) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java:745]: remove extra new line

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestSecureOzoneCluster.java (hunk 3 above)
Comment (sadanand48) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestSecureOzoneCluster.java:1417]: +1

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/S3SecretManagerImpl.java [hunk 11]
```diff
@@ -118,4 +129,11 @@ public S3SecretCache cache() {
   public S3Batcher batcher() {
     return s3SecretStore.batcher();
   }
+
+  public void updateCache(String kerberosID, S3SecretValue secret) {
+    S3SecretManager.super.updateCache(kerberosID, secret);
+  }
+
+
+
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/S3SecretManagerImpl.java:138]: remove extra new lines

**Review by sadanand48 — COMMENTED** [review]
Thanks @ArafatKhan2198 for continuous updates to the patch, Overall patch looks good. We should add some more unit tests for the S3 cache behaviour that this patch adds. We can also do it in a separate change instead of blocking this one.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java (hunk 7 above)
Comment (ChenSammi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java:208]: This is not addressed yet @ArafatKhan2198 .

**Comment by ChenSammi** [comment]
@ArafatKhan2198 , could you add a new test which calls S3SecretCache#clearCache, check that the cache size is reduced after the call?

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerDoubleBuffer.java [hunk 12]
```diff
@@ -18,9 +18,7 @@
 
 import java.io.File;
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
+import java.util.*;
```
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerDoubleBuffer.java:21]: Please don't use * in import.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerDoubleBuffer.java [hunk 13]
```diff
@@ -30,27 +28,39 @@
 import org.apache.hadoop.hdds.conf.OzoneConfiguration;
 import org.apache.hadoop.ozone.audit.AuditLogger;
 import org.apache.hadoop.ozone.audit.AuditMessage;
-import org.apache.hadoop.ozone.om.OMConfigKeys;
-import org.apache.hadoop.ozone.om.OMMetadataManager;
-import org.apache.hadoop.ozone.om.OMMetrics;
-import org.apache.hadoop.ozone.om.OmMetadataManagerImpl;
-import org.apache.hadoop.ozone.om.OzoneManager;
+import org.apache.hadoop.ozone.om.*;
```
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerDoubleBuffer.java:31]: Please don't use * in import.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/s3/security/TestS3GetSecretRequest.java [hunk 14]
```diff
@@ -24,29 +24,25 @@
...
 import org.apache.hadoop.ozone.audit.AuditMessage;
-import org.apache.hadoop.ozone.om.OMConfigKeys;
-import org.apache.hadoop.ozone.om.OMMetrics;
-import org.apache.hadoop.ozone.om.OMMultiTenantManager;
-import org.apache.hadoop.ozone.om.OmMetadataManagerImpl;
-import org.apache.hadoop.ozone.om.OzoneManager;
-import org.apache.hadoop.ozone.om.S3SecretLockedManager;
-import org.apache.hadoop.ozone.om.S3SecretManagerImpl;
-import org.apache.hadoop.ozone.om.TenantOp;
+import org.apache.hadoop.ozone.om.*;
```
Comment (ChenSammi) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/s3/security/TestS3GetSecretRequest.java:27]: Please don't use * in import.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerDoubleBuffer.java (hunk 12 above)
Comment (adoroszlai) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerDoubleBuffer.java:21]: Checkstyle agrees ;)

https://github.com/ArafatKhan2198/ozone/actions/runs/6920030792/job/18824061794#step:7:12

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestSecureOzoneCluster.java (hunk 3 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestSecureOzoneCluster.java:1447]: Hi @ChenSammi I am not sure why this is coming as I dont see any local changes that I have made to this class 
image

**Review by ChenSammi — APPROVED** [review]

**Review by sadanand48 — APPROVED** [review]

**Comment by adoroszlai** [comment]
Test passed in [10x25 run](https://github.com/adoroszlai/ozone/actions/runs/6949345205) with latest patch.

Thanks @ArafatKhan2198 for the patch, @ChenSammi, @Pochatkin, @sadanand48, @tanvipenumudy for the review.
